app/models/site.rb

```diff
@@ -28,16 +28,41 @@ def user_fields
     UserField.order(:position).all
   end
 
-  def categories
-    @categories ||= begin
+  CATEGORIES_CACHE_KEY = "site_categories"
+
+  def self.clear_cache
+    Discourse.cache.delete(CATEGORIES_CACHE_KEY)
+  end
+
+  def self.all_categories_cache
+    # Categories do not change often so there is no need for us to run the
+    # same query and spend time creating ActiveRecord objects for every requests.
+    #
+    # Do note that any new association added to the eager loading needs a
+    # corresponding ActiveRecord callback to clear the categories cache.
+    Discourse.cache.fetch(CATEGORIES_CACHE_KEY, expires_in: 30.minutes) do
       categories = Category
-        .includes(:uploaded_logo, :uploaded_background, :tags, :tag_groups)
-        .secured(@guardian)
+        .includes(:uploaded_logo, :uploaded_background, :tags, :tag_groups, :required_tag_group)
         .joins('LEFT JOIN topics t on t.id = categories.topic_id')
         .select('categories.*, t.slug topic_slug')
         .order(:position)
+        .to_a
+
+      ActiveModel::ArraySerializer.new(
+        categories,
+        each_serializer: SiteCategorySerializer
+      ).as_json
+    end
+  end
+
+  def categories
+    @categories ||= begin
+      categories = []
 
-      categories = categories.to_a
+      self.class.all_categories_cache.each do |category|
+        c = OpenStruct.new(category)
```

Comment:
this feels somewhat fragile.

https://github.com/discourse/discourse/blob/b8a29800697b38dabae28b0c36590647b84ffada/lib/guardian/category_guardian.rb#L49-L57

instead I would define a special guardian that `can_see_category` calls.

maybe can_see_serialized_category(category_id, read_restricted) ... not sure but the open struct there is not deliberate enough. 

staged user check can be bypassed on the serialized form and maybe can_see_category calls it?